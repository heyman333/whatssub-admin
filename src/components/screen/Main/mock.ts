import faker from 'faker';
import { Service } from '../../../types';

// This function is being used to create below SERVICE_LIST

// const createFakeServiceList = () => {
//   let list: Array<Service> = [];
//   const length = 50;
//   for (let i = 0; i < length; i++) {
//     list.push({
//       id: faker.random.number().toString(),
//       name: faker.company.companyName(),
//       nameKr: faker.company.companyName(),
//       icon: faker.image.avatar(),
//     });
//   };
//   // you can copy the result of it and paste it to 'ServiceList'
//   console.log('fake service list', JSON.stringify(list, null, 2));
//   return list;
// };

export const SERVICE_LIST: Array<Service> = [
  {
    'id': '66326',
    'name': 'Dickinson LLC',
    'nameKr': 'Prosacco, Schowalter and Weissnat',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/kylefrost/128.jpg',
  },
  {
    'id': '96883',
    'name': 'Little, Fritsch and Schmitt',
    'nameKr': 'Williamson, Weissnat and Prosacco',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/justinrob/128.jpg',
  },
  {
    'id': '69533',
    'name': 'Franecki - Erdman',
    'nameKr': 'Bogisich and Sons',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/madewulf/128.jpg',
  },
  {
    'id': '64598',
    'name': 'Schinner LLC',
    'nameKr': 'Carroll, Walter and Heathcote',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/dahparra/128.jpg',
  },
  {
    'id': '1805',
    'name': 'Kunde - Zulauf',
    'nameKr': 'Renner, Moore and Williamson',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/stevedesigner/128.jpg',
  },
  {
    'id': '67807',
    'name': 'Maggio Inc',
    'nameKr': 'Zemlak and Sons',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/bboy1895/128.jpg',
  },
  {
    'id': '74063',
    'name': 'Kozey, Turcotte and Jerde',
    'nameKr': 'Kuhlman and Sons',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/adellecharles/128.jpg',
  },
  {
    'id': '4394',
    'name': 'Dickens - Tromp',
    'nameKr': 'Becker, Nolan and Miller',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/andrea211087/128.jpg',
  },
  {
    'id': '98903',
    'name': 'Lemke - Sipes',
    'nameKr': 'Price, Sipes and Cormier',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/peejfancher/128.jpg',
  },
  {
    'id': '67264',
    'name': 'Koepp - Sporer',
    'nameKr': 'Reichel, Christiansen and Schneider',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/mrjamesnoble/128.jpg',
  },
  {
    'id': '71896',
    'name': 'Hintz Group',
    'nameKr': 'Rippin LLC',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/saarabpreet/128.jpg',
  },
  {
    'id': '64305',
    'name': 'Koch, Thiel and Vandervort',
    'nameKr': 'Hoeger, Beatty and Kessler',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/moscoz/128.jpg',
  },
  {
    'id': '41401',
    'name': 'Hoppe, Stoltenberg and Cremin',
    'nameKr': 'Pfeffer Inc',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/gipsy_raf/128.jpg',
  },
  {
    'id': '52945',
    'name': 'Monahan - Hintz',
    'nameKr': 'Beer LLC',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/laurengray/128.jpg',
  },
  {
    'id': '73657',
    'name': 'Kub, Doyle and Halvorson',
    'nameKr': 'Osinski, Medhurst and Cronin',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/shesgared/128.jpg',
  },
  {
    'id': '72388',
    'name': 'Sawayn - Paucek',
    'nameKr': 'Paucek, Kuhn and Monahan',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/uxalex/128.jpg',
  },
  {
    'id': '36458',
    'name': 'Mayert, Greenfelder and Mills',
    'nameKr': 'Botsford, Emmerich and Heller',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/mattbilotti/128.jpg',
  },
  {
    'id': '93178',
    'name': 'Mohr - Marquardt',
    'nameKr': 'Langosh Inc',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/davecraige/128.jpg',
  },
  {
    'id': '29400',
    'name': 'Johnson Inc',
    'nameKr': 'Langosh, Kris and Quitzon',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/praveen_vijaya/128.jpg',
  },
  {
    'id': '8636',
    'name': 'Schroeder Inc',
    'nameKr': 'Volkman, Towne and Roob',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/VinThomas/128.jpg',
  },
  {
    'id': '18768',
    'name': 'Franecki LLC',
    'nameKr': 'Boehm - Vandervort',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/bublienko/128.jpg',
  },
  {
    'id': '1302',
    'name': 'Ferry, Schoen and Corwin',
    'nameKr': 'Mraz, Lowe and Lindgren',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/calebjoyce/128.jpg',
  },
  {
    'id': '15156',
    'name': 'Morar, Vandervort and Stark',
    'nameKr': "O'Conner Inc",
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/betraydan/128.jpg',
  },
  {
    'id': '5459',
    'name': 'Bartell - Herzog',
    'nameKr': 'Effertz, Cassin and Robel',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/cadikkara/128.jpg',
  },
  {
    'id': '17392',
    'name': 'Littel Inc',
    'nameKr': 'Prohaska LLC',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/h1brd/128.jpg',
  },
  {
    'id': '99635',
    'name': 'Jacobson and Sons',
    'nameKr': 'Hoeger - Daniel',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/thatonetommy/128.jpg',
  },
  {
    'id': '67587',
    'name': 'Reichert, Cole and Ratke',
    'nameKr': 'Rippin - Goyette',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/krasnoukhov/128.jpg',
  },
  {
    'id': '3266',
    'name': 'Marquardt - Denesik',
    'nameKr': 'Johnson, Streich and Gleason',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/geshan/128.jpg',
  },
  {
    'id': '29658',
    'name': 'Orn and Sons',
    'nameKr': 'Bernier Group',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/abdullindenis/128.jpg',
  },
  {
    'id': '48531',
    'name': 'Moen, Abbott and Gleason',
    'nameKr': 'Purdy - Durgan',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/suribbles/128.jpg',
  },
  {
    'id': '99395',
    'name': 'Corkery - Kulas',
    'nameKr': 'Satterfield, Heller and Fisher',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/kerihenare/128.jpg',
  },
  {
    'id': '83962',
    'name': 'Kiehn - Roob',
    'nameKr': 'Douglas Group',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/oscarowusu/128.jpg',
  },
  {
    'id': '60683',
    'name': 'Orn - Wiegand',
    'nameKr': "Hintz - O'Keefe",
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/derekcramer/128.jpg',
  },
  {
    'id': '96307',
    'name': 'Mills - Kuvalis',
    'nameKr': 'DuBuque - Dibbert',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/mwarkentin/128.jpg',
  },
  {
    'id': '56053',
    'name': 'Gerhold - Bartell',
    'nameKr': 'Feil - Kling',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/Skyhartman/128.jpg',
  },
  {
    'id': '40807',
    'name': 'Mitchell, Osinski and Weissnat',
    'nameKr': 'Schulist - Kozey',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/peachananr/128.jpg',
  },
  {
    'id': '19355',
    'name': 'Price - Zboncak',
    'nameKr': 'Becker, Corkery and Maggio',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/dicesales/128.jpg',
  },
  {
    'id': '27582',
    'name': 'McClure Group',
    'nameKr': 'Lowe, Windler and Dooley',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/bradenhamm/128.jpg',
  },
  {
    'id': '70457',
    'name': 'Wyman LLC',
    'nameKr': 'Lemke, Kub and Baumbach',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/gregkilian/128.jpg',
  },
  {
    'id': '90123',
    'name': 'Moen Group',
    'nameKr': 'Hayes - Von',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/eduardostuart/128.jpg',
  },
  {
    'id': '51747',
    'name': 'Ondricka, Smith and Abbott',
    'nameKr': 'Nader and Sons',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/vickyshits/128.jpg',
  },
  {
    'id': '87355',
    'name': 'Davis, Howell and Lindgren',
    'nameKr': 'Fahey - Bradtke',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/primozcigler/128.jpg',
  },
  {
    'id': '61474',
    'name': 'Mitchell LLC',
    'nameKr': 'Buckridge Group',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/ryandownie/128.jpg',
  },
  {
    'id': '35247',
    'name': 'MacGyver - Thompson',
    'nameKr': 'Blick - Hilll',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/magugzbrand2d/128.jpg',
  },
  {
    'id': '30366',
    'name': 'Waters - Lubowitz',
    'nameKr': 'Mills - Kohler',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/charliecwaite/128.jpg',
  },
  {
    'id': '81822',
    'name': 'Daniel - Friesen',
    'nameKr': 'Batz and Sons',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/stevedesigner/128.jpg',
  },
  {
    'id': '13399',
    'name': 'Wilderman - Hoppe',
    'nameKr': 'Stroman, Emmerich and Howell',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/brajeshwar/128.jpg',
  },
  {
    'id': '83969',
    'name': 'Lubowitz - Monahan',
    'nameKr': 'Grimes - Quigley',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/antjanus/128.jpg',
  },
  {
    'id': '61979',
    'name': 'Lemke, Bergstrom and Hane',
    'nameKr': 'Kihn LLC',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/heycamtaylor/128.jpg',
  },
  {
    'id': '99082',
    'name': 'Cummings LLC',
    'nameKr': 'Schuster and Sons',
    'icon': 'https://s3.amazonaws.com/uifaces/faces/twitter/stan/128.jpg',
  },
];
