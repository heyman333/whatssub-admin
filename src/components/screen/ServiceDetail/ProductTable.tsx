import React, { useState } from 'react';
import { IProduct } from '../../../types';
import styled, { css } from 'styled-components';
import { getString } from '../../../../STRINGS';
import { colors } from '../../../theme';
import SimpleButton from '../../shared/SimpleButton';

const productLabel: IProduct =
{
  id: 'ID',
  name: '상품명',
  currency: '통화',
  price: '가격',
  subType: '갱신단위 / 주기',
};

const Hide = css`
  position: absolute;
  top: -9999px;
  left: -9999px;
`;

const WhenNarrow = `
  @media only screen and (max-width: 760px),
  (min-device-width: 768px) and (max-device-width: 1024px)
`;

const cellPadding = css`
  padding: 0 0 0 0.5em;
`;

const Container = styled.div`
  overflow-y: auto; 
  height: 100%;
  width: 100%;
`;

const Table = styled.table`
  margin: auto;

  width: 95%;
  ${WhenNarrow} {
    display: block;
  }
`;

const Thead = styled.thead`
  ${WhenNarrow} {
    ${Hide}
  }
`;

const Th = styled.th`
  z-index: 1;
  position: sticky; 
  top: 0;
  background:#ccc;
  color:${colors.darkBackgroundLight};

  ${cellPadding}
`;

const Tr = styled.tr`
  color: ${({ theme }) => theme.fontColor};
  background-color: ${({ theme }) => theme.background};
  ${WhenNarrow} {
    display: block;  
    margin: 0 0 1rem 0;    
  }
  &:nth-child(even){
    background-color: #eee;
    color: ${colors.darkBackgroundLight};
  }
  &:hover{
    background-color: ${({ theme }) => theme.serviceListHoveColor};
  }
`;

const Tbody = styled.tbody`
  ${WhenNarrow} {
    display: block;
  }
`;

const leftSideOverRightSidePortion = `35%`;
const labelMarginLift = `0.2em`;
const labelContentWhenNarrow = css`
  /* like a table header */
  position: absolute;
  /* Top/left values mimic padding */
  top: 0;
  left: 0;

  margin-left: ${labelMarginLift};
  width: ${leftSideOverRightSidePortion}-${labelMarginLift};
  white-space: nowrap;
`;
interface ICell {
  label: string
}
const Cell = styled.td<ICell>`
  text-align: center;
  vertical-align: middle;
  ${cellPadding}

  ${WhenNarrow} {
    display: block;
    /* like a row */
    border: none;
    border-bottom: 1px solid #eee;
    position: relative;
    padding-left: ${leftSideOverRightSidePortion};

    text-align: left;
    &:before {
      ${labelContentWhenNarrow}
      content: '${(props) => props.label}';
    }
  }
`;

interface IClickableCell {
  scaleUp?: boolean
}
const ClickableCell = styled(Cell) <IClickableCell>`
  &:hover {
    cursor: pointer;
  }

  /* Visual feedback*/
  transition: all .1s ease-in-out;
  transform: ${(props) => props.scaleUp ? `scale(1.1)` : `none`};
  ${WhenNarrow} {
    transform: scale(1);
  }
`;

const ControlCell = styled(Cell)`
  /* centering buttons */
  vertical-align: middle;
  height:42px;
  ${WhenNarrow}{
    height: auto;
  }

  display: flex;
  justify-content: center;
  align-items: center;  
`;

interface IProps {
  onDeleteProductClick?: (productId: string) => void;
  onProductClick?: (productId: string) => void;
  productList: Array<IProduct>;
}
const ProductTable: React.FC<IProps> = (props) => {
  const [hoverRow, setHoverRow] = useState('');
  const content = props.productList.map(({ id, name, currency, price, subType }) => {
    // scale up some cells when mouse is hovering on its parent's row -> visual feedback for user
    const scaleUp = id === hoverRow;
    return (
      <Tr data-testid={`tableRow-${id}`} onMouseOver={() => { setHoverRow(id); }} key={id} onClick={(event: any) => {
        // Using bubbling events from clicking desired cells to fire navigation
        // Currently event condition fires with all cells execpt ControlCell. in ControlCell it uses `event.stopPropagation()` to prevent event to bubble up
        props.onProductClick && props.onProductClick(id);
        // TODO: Integrate 'ProductDetail' screen navigation. remove below 'console.log()' afterwards
        // console.log(`Navigate to 'ProductDetail' Screen with Product ID:${id}`);
      }
      }>
        <ClickableCell data-testid={`idCell-${id}`} label={productLabel.id} scaleUp={scaleUp}>{id}</ClickableCell>
        <ClickableCell data-testid={`nameCell-${id}`} label={productLabel.name} scaleUp={scaleUp}>{name}</ClickableCell>
        <ClickableCell data-testid={`currencyCell-${id}`} label={productLabel.currency} scaleUp={scaleUp}>{currency}</ClickableCell>
        <ClickableCell data-testid={`priceCell-${id}`} label={productLabel.price} scaleUp={scaleUp}>{price}</ClickableCell>
        <ClickableCell data-testid={`subTypeCell-${id}`} label={productLabel.subType} scaleUp={scaleUp}>{subType}</ClickableCell>

        <ControlCell data-testid={`controlCell-${id}`} label={null} onClick={(e) => { e.stopPropagation(); }}>
          <SimpleButton data-testid={`deleteProductButton-${id}`}
            onClick={() => {
              props.onDeleteProductClick && props.onDeleteProductClick(id);
              // TODO: Integrate opening 'ProductModalDelete' confirmation modal. remove below 'console.log()' afterwards
              // console.log(`Delete button is clicked. Open confirming modal for deleting Product Id:${id}`);
            }}>{getString('DELETE_BUTTON')}</SimpleButton>
        </ControlCell>
      </Tr >
    );
  });
  return (
    <Container>
      <Table>
        <Thead>
          <Tr>
            <Th>{productLabel.id}</Th>
            <Th>{productLabel.name}</Th>
            <Th>{productLabel.currency}</Th>
            <Th>{productLabel.price}</Th>
            <Th>{productLabel.subType}</Th>
            <Th>{null}</Th>
          </Tr>
        </Thead>
        <Tbody >
          {content}
        </Tbody>
      </Table>
    </Container>
  );
};

export default ProductTable;
