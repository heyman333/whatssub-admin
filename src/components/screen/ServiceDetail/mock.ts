import faker from 'faker';
import { IProduct, ServiceForServiceDetail } from '../../../types';

export const PRODUCT_LIST: Array<IProduct> = [{
  'id': '1',
  'name': 'Basic',
  'currency': 'KRW',
  'price': '5,000',
  'subType': '월 / 1',
},
{
  'id': '2',
  'name': 'Pro',
  'currency': 'KRW',
  'price': '120,000',
  'subType': '년 / 1',
},
{
  'id': '3',
  'name': 'Special',
  'currency': 'USD',
  'price': '110.00',
  'subType': '년 / 1',
}];

export const SERVICE_DETAIL: ServiceForServiceDetail = {
  name: 'Netflix',
  nameKr: '넷플릭스',
  url: 'netflix.com',
  description: '넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. ',
};
