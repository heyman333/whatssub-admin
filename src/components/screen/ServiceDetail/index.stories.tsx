import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { linkTo } from '@storybook/addon-links';
import ServiceDetail from '.';
import { PRODUCT_LIST, SERVICE_DETAIL } from './mock';

const Story = () => (
  <ServiceDetail
    productList={PRODUCT_LIST}
    serviceDetail={SERVICE_DETAIL}
    onAddProductClick={action('onAddProduct')}
    onDeleteProductClick={action('onDeleteProduct')}
    onProductClick={action('onProduct')} />);

storiesOf('ServiceDetail', module)
  .add('default', () => <Story />)
  .add('iphonex', () => <Story />, { viewport: { defaultViewport: 'iphonex' } });
