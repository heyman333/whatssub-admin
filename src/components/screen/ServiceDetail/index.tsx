import React, { Component } from 'react';
import styled from 'styled-components';
import ProductTable from './ProductTable';
import { getString } from '../../../../STRINGS';
import SimpleButton from '../../shared/SimpleButton';
import { IProduct, ServiceForServiceDetail } from '../../../types';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  background: ${(props) => props.theme.background};
  color: ${(props) => props.theme.fontColor};
`;

const Title = styled.h1`
  flex: 0 0 auto;
  align-self: flex-start;
  font-size: 2em;
  padding: 0.5em 5%;
`;

const Detail = styled.div`
  margin: 0.3em 5%;
  align-self: flex-start;
  display: flex;
  justify-content: flex-start;
`;
const Label = styled.div`
  font-weight: 800;
  flex: 0 0 5em;
`;

const ProductListTitle = styled.div`
  text-align: left;
`;

const TableWrapper = styled.div`
  align-self: center;
  width: 100%;
  max-width: 1100px;

  display: flex;
  flex-direction: column;
  align-items: center;
`;

const TableLabelAndControl = styled.div`
  flex: 0 0 auto;
  display: flex;
  justify-content: space-between;
  width: 90%;
  margin: 1em 1% 0.5em 1%;
`;

const AddProductButton = styled(SimpleButton)`
  margin-right: 0.5em;
`;

interface IServiceDetail {
  history?: any;
  // Those Product related click handlers are used in test
  onAddProductClick?: () => void;
  onProductClick?: (productId: string) => void;
  onDeleteProductClick?: (productId: string) => void;
  serviceDetail: ServiceForServiceDetail;
  productList: Array<IProduct>;
}

function ServiceDetail(props: IServiceDetail) {
  const { history, onAddProductClick, serviceDetail, ...propsToProductTable } = props;
  const { name, url, description } = serviceDetail;
  return (
    <Container>
      <TableWrapper>
        <Title>Whatssub Admin</Title>
        <Detail>
          <Label>서비스 명</Label>
          <div>{name}</div>
        </Detail>
        <Detail>
          <Label>URL</Label>
          <div>{url}</div>
        </Detail>
        <Detail>
          <Label>설명</Label>
          <div>{description}</div>
        </Detail>

        <TableLabelAndControl>
          <ProductListTitle>{getString('PRODUCT_LIST')}</ProductListTitle>
          <AddProductButton data-testid='addProductButton' onClick={() => {
            props.onAddProductClick && props.onAddProductClick();
            // TODO: Integrate opening 'ProductModalEdit'. remove below 'console.log()' afterwards
            // console.log(`Add Product button is clicked. Open 'ProductModalEdit'`);
          }}>{getString('PRODUCT_ADD_BUTTON')}</AddProductButton>
        </TableLabelAndControl>
        <ProductTable {...propsToProductTable} />
      </TableWrapper>
    </Container>
  );
}

export default ServiceDetail;
