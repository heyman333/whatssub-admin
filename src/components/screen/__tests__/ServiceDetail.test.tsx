import React from 'react';
import * as renderer from 'react-test-renderer';

import ServiceDetail from '../ServiceDetail';
import { render, fireEvent, getByTestId, wait, queryByTestId } from '@testing-library/react';
import { PRODUCT_LIST, SERVICE_DETAIL } from '../ServiceDetail/mock';

const props = {
  onAddProductClick: jest.fn(),
  onProductClick: jest.fn(),
  onDeleteProductClick: jest.fn(),
  serviceDetail: SERVICE_DETAIL,
  productList: PRODUCT_LIST,
};

describe('[ServiceDetail] render', () => {
  it('renders without crashing', () => {
    const rendered = renderer.create(<ServiceDetail {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[ServiceDetail] Interaction', () => {
  const component = <ServiceDetail {...props} />;
  let renderResult: any;

  beforeAll(() => {
    renderResult = render(component);
  });

  it('simulate [onAddProductClick] is called when corresponding button is clicked', () => {
    const addProductButton = renderResult.getByTestId('addProductButton');
    fireEvent.click(addProductButton);
    expect(props.onAddProductClick).toHaveBeenCalled();
  });

  it('simulate [onDeleteProductClick] with desired [product id] as parameter is called when corresponding buttons is clicked', () => {
    PRODUCT_LIST.map((product, index) => {
      const chosenProductId = product.id;
      const deleteProductButton = renderResult.getByTestId(`deleteProductButton-${chosenProductId}`);
      fireEvent.click(deleteProductButton);
      expect(props.onDeleteProductClick.mock.calls[index][0]).toBe(chosenProductId);
    });
  });

  describe('simulate [onProductClick] with corresponding [product id]. and it should fire when clicking its specific children cells', () => {
    beforeEach(() => {
      // clear called counts
      props.onProductClick.mockClear();
    });

    it('simulate clicking desired children cells call [onProductClick]', () => {
      const testIdPrefixOfCells = ['idCell-', 'nameCell-', 'currencyCell-', 'priceCell-', 'subTypeCell-'];
      let callCount = 0;
      PRODUCT_LIST.map(({ id }) => {
        testIdPrefixOfCells.map((prefix) => {
          const cell = renderResult.getByTestId(`${prefix}${id}`);
          fireEvent.click(cell);
          expect(props.onProductClick.mock.calls[callCount++][0]).toBe(id);
        });
      });
    });

    it('simulate clicking control cell doesnt call [onProductClick]', () => {
      const testIdPrefixOfCells = ['controlCell-'];
      PRODUCT_LIST.map(({ id }) => {
        testIdPrefixOfCells.map((prefix) => {
          const cell = renderResult.getByTestId(`${prefix}${id}`);
          fireEvent.click(cell);
          expect(props.onProductClick).toHaveBeenCalledTimes(0);
        });
      });
    });
  });
});
