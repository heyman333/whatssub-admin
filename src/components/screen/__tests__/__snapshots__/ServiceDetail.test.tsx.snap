// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ServiceDetail] render renders without crashing 1`] = `
<div
  className="sc-iwsKbI bXavdl"
>
  <div
    className="sc-fjdhpX ewNMeA"
  >
    <h1
      className="sc-gZMcBi dBxpMD"
    >
      Whatssub Admin
    </h1>
    <div
      className="sc-gqjmRU jhdAMm"
    >
      <div
        className="sc-VigVT iEQGci"
      >
        서비스 명
      </div>
      <div>
        Netflix
      </div>
    </div>
    <div
      className="sc-gqjmRU jhdAMm"
    >
      <div
        className="sc-VigVT iEQGci"
      >
        URL
      </div>
      <div>
        netflix.com
      </div>
    </div>
    <div
      className="sc-gqjmRU jhdAMm"
    >
      <div
        className="sc-VigVT iEQGci"
      >
        설명
      </div>
      <div>
        넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 넷플릭스 입니다. 
      </div>
    </div>
    <div
      className="sc-jzJRlG dVXBlX"
    >
      <div
        className="sc-jTzLTM fSJIUY"
      >
        Product List
      </div>
      <button
        className="sc-bdVaJa sc-cSHVUG kktCio"
        data-testid="addProductButton"
        onClick={[Function]}
      >
        Add Product
      </button>
    </div>
    <div
      className="sc-bwzfXH fIDMNA"
    >
      <table
        className="sc-htpNat jbGzjU"
      >
        <thead
          className="sc-bxivhb ldJWmR"
        >
          <tr
            className="sc-EHOje fIbXKp"
          >
            <th
              className="sc-ifAKCX ftnFMk"
            >
              ID
            </th>
            <th
              className="sc-ifAKCX ftnFMk"
            >
              상품명
            </th>
            <th
              className="sc-ifAKCX ftnFMk"
            >
              통화
            </th>
            <th
              className="sc-ifAKCX ftnFMk"
            >
              가격
            </th>
            <th
              className="sc-ifAKCX ftnFMk"
            >
              갱신단위 / 주기
            </th>
            <th
              className="sc-ifAKCX ftnFMk"
            />
          </tr>
        </thead>
        <tbody
          className="sc-bZQynM caJKpL"
        >
          <tr
            className="sc-EHOje fIbXKp"
            data-testid="tableRow-1"
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            <td
              className="sc-gzVnrw sc-htoDjs jWoYzQ"
              data-testid="idCell-1"
              label="ID"
            >
              1
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs jYKhlX"
              data-testid="nameCell-1"
              label="상품명"
            >
              Basic
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs pxokY"
              data-testid="currencyCell-1"
              label="통화"
            >
              KRW
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs bhArpJ"
              data-testid="priceCell-1"
              label="가격"
            >
              5,000
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs exOXOa"
              data-testid="subTypeCell-1"
              label="갱신단위 / 주기"
            >
              월 / 1
            </td>
            <td
              className="sc-gzVnrw sc-dnqmqq dqYWns"
              data-testid="controlCell-1"
              label={null}
              onClick={[Function]}
            >
              <button
                className="sc-bdVaJa kLNRSS"
                data-testid="deleteProductButton-1"
                onClick={[Function]}
              >
                Delete
              </button>
            </td>
          </tr>
          <tr
            className="sc-EHOje fIbXKp"
            data-testid="tableRow-2"
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            <td
              className="sc-gzVnrw sc-htoDjs jWoYzQ"
              data-testid="idCell-2"
              label="ID"
            >
              2
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs jYKhlX"
              data-testid="nameCell-2"
              label="상품명"
            >
              Pro
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs pxokY"
              data-testid="currencyCell-2"
              label="통화"
            >
              KRW
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs bhArpJ"
              data-testid="priceCell-2"
              label="가격"
            >
              120,000
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs exOXOa"
              data-testid="subTypeCell-2"
              label="갱신단위 / 주기"
            >
              년 / 1
            </td>
            <td
              className="sc-gzVnrw sc-dnqmqq dqYWns"
              data-testid="controlCell-2"
              label={null}
              onClick={[Function]}
            >
              <button
                className="sc-bdVaJa kLNRSS"
                data-testid="deleteProductButton-2"
                onClick={[Function]}
              >
                Delete
              </button>
            </td>
          </tr>
          <tr
            className="sc-EHOje fIbXKp"
            data-testid="tableRow-3"
            onClick={[Function]}
            onMouseOver={[Function]}
          >
            <td
              className="sc-gzVnrw sc-htoDjs jWoYzQ"
              data-testid="idCell-3"
              label="ID"
            >
              3
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs jYKhlX"
              data-testid="nameCell-3"
              label="상품명"
            >
              Special
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs pxokY"
              data-testid="currencyCell-3"
              label="통화"
            >
              USD
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs bhArpJ"
              data-testid="priceCell-3"
              label="가격"
            >
              110.00
            </td>
            <td
              className="sc-gzVnrw sc-htoDjs exOXOa"
              data-testid="subTypeCell-3"
              label="갱신단위 / 주기"
            >
              년 / 1
            </td>
            <td
              className="sc-gzVnrw sc-dnqmqq dqYWns"
              data-testid="controlCell-3"
              label={null}
              onClick={[Function]}
            >
              <button
                className="sc-bdVaJa kLNRSS"
                data-testid="deleteProductButton-3"
                onClick={[Function]}
              >
                Delete
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
